File: app/modules/gemini_usage/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class GeminiUsageWindowResponse(DashboardModel):
    window: str
    label: str
    remaining_percent: float = Field(ge=0, le=100)
    used_percent: float = Field(ge=0, le=100)
    resets_at: datetime
    captured_at: datetime


class GeminiUsageMonitorResponse(DashboardModel):
    configured: bool
    last_attempt_at: datetime | None = None
    last_success_at: datetime | None = None
    last_error: str | None = None
    windows: list[GeminiUsageWindowResponse] = Field(default_factory=list)


class GeminiUsageCredentialRequest(DashboardModel):
    refresh_token: str = Field(min_length=1, max_length=4096)

File: frontend/src/features/gemini-usage/schemas.ts
import { z } from "zod";

export const GeminiUsageWindowSchema = z.object({
  window: z.string(), label: z.string(), remainingPercent: z.number().min(0).max(100), usedPercent: z.number().min(0).max(100),
  resetsAt: z.string().datetime({ offset: true }), capturedAt: z.string().datetime({ offset: true }),
});
export const GeminiUsageMonitorSchema = z.object({
  configured: z.boolean(), lastAttemptAt: z.string().datetime({ offset: true }).nullable(), lastSuccessAt: z.string().datetime({ offset: true }).nullable(), lastError: z.string().nullable(), windows: z.array(GeminiUsageWindowSchema),
});
export const GeminiUsageCredentialRequestSchema = z.object({ refreshToken: z.string().trim().min(1) });
export type GeminiUsageMonitor = z.infer<typeof GeminiUsageMonitorSchema>;
